cluster-autoscaler/simulator/scheduler_based_predicates_checker.go

```diff
@@ -0,0 +1,339 @@
+		return nil, fmt.Errorf("couldn't create scheduler framework; %v", err)
+	}
+
+	checker := &SchedulerBasedPredicateChecker{
+		framework:                framework,
+		delegatingSharedLister:   sharedLister,
+		informerBasedShardLister: NewEmptySnapshot(),
+		nodeLister:               informerFactory.Core().V1().Nodes().Lister(),
+		podLister:                informerFactory.Core().V1().Pods().Lister(),
+	}
+
+	// this MUST be called after all the informers/listers are acquired via the
+	// informerFactory....Lister()/informerFactory....Informer() methods
+	informerFactory.Start(stop)
+
+	return checker, nil
+}
+
+// SnapshotClusterState updates cluster snapshot used by the predicate checker.
+// It should be called every CA loop iteration.
+func (p *SchedulerBasedPredicateChecker) SnapshotClusterState() error {
+	nodes, err := p.nodeLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("could not list Nodes; %v", err)
+	}
+	pods, err := p.podLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("could not list Pods; %v", err)
+	}
+	nodeInfoMap := CreateNodeInfoMap(pods, nodes)
+	newSnapshot := NewSnapshot(nodeInfoMap)
+	p.delegatingSharedLister.UpdateDelegate(newSnapshot)
+	return nil
+}
+
+// FitsAnyNode checks if the given pod can be place on any of the given nodes.
+func (p *SchedulerBasedPredicateChecker) FitsAnyNode(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod, nodeInfos map[string]*scheduler_nodeinfo.NodeInfo) (string, error) {
+	if clusterSnapshot != nil {
+		if nodeInfos != nil {
+			klog.Errorf("clusterSnapshot and nodeInfos are mutually exclusive!!!!")
+		}
+		return p.fitsAnyNode(clusterSnapshot, pod)
+	}
+	return p.fitsAnyNodeDeprecated(pod, nodeInfos)
+}
+
+// FitsAnyNode checks if the given pod can be place on any of the given nodes.
+func (p *SchedulerBasedPredicateChecker) fitsAnyNodeDeprecated(pod *apiv1.Pod, nodeInfos map[string]*scheduler_nodeinfo.NodeInfo) (string, error) {
+	p.delegatingSharedLister.UpdateDelegate(p.informerBasedShardLister)
+	defer p.delegatingSharedLister.ResetDelegate()
+
+	state := scheduler_framework.NewCycleState()
+	preFilterStatus := p.framework.RunPreFilterPlugins(context.TODO(), state, pod)
+	if !preFilterStatus.IsSuccess() {
+		return "", fmt.Errorf("error running pre filter plugins for pod %s; %s", pod.Name, preFilterStatus.Message())
+	}
+
+	for _, nodeInfo := range nodeInfos {
+		// Be sure that the node is schedulable.
+		if nodeInfo.Node().Spec.Unschedulable {
+			continue
+		}
+
+		filterStatuses := p.framework.RunFilterPlugins(context.TODO(), state, pod, nodeInfo)
+		ok := true
+		for _, filterStatus := range filterStatuses {
+			if !filterStatus.IsSuccess() {
+				ok = false
+				break
+			}
+		}
+		if ok {
+			return nodeInfo.Node().Name, nil
+		}
+	}
+	return "", fmt.Errorf("cannot put pod %s on any node", pod.Name)
+}
+
+func (p *SchedulerBasedPredicateChecker) fitsAnyNode(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod) (string, error) {
```

Comment:
As commented earlier - this should probably take a subset of nodes to consider.

Reply:
I buy the prefiltering argument. Please add this to followup list

Reply:
One thing that would need to be addressed then is the returning of actual predicate errors from the coll (map node->error may be needed)

Reply:
I'm pretty sure we ignore specific errors in binpacking anyway.
Added to follow-up list.